File: flutter/packages/hatake_core/lib/src/tooling/page_kind_words.dart
import 'page_kinds.dart';

/// How a page kind is said in Japanese.
///
/// Two wordings per kind on purpose: [what] is the sentence an explanation uses
/// ("検索して一覧に出し、その場で登録・修正・削除までできる画面"), [short] is the
/// heading word a one-line summary or an index needs ("検索＋一覧＋登録・修正・削除").
/// A sentence does not fit in a table row, and a heading word does not explain.
class PageKindWords {
  const PageKindWords({required this.what, required this.short});

  /// Full wording — what this screen is for.
  final String what;

  /// Heading word — fits one line of a table.
  final String short;
}

/// Page kind → wording, transcribed from `spec/vocabulary.json`.
///
/// The spec file is the source of truth (the TypeScript and Java editions
/// transcribe the same table), so the wording cannot drift between the CLI and
/// what an app shows. `page_kind_words_test.dart` compares this table with the
/// spec key by key, which is why transcribing by hand is safe here.
const Map<String, PageKindWords> pageKindWords = {
  PageKinds.crud: PageKindWords(
    what: '検索して一覧に出し、その場で登録・修正・削除までできる画面',
    short: '検索＋一覧＋登録・修正・削除',
  ),
  PageKinds.master: PageKindWords(
    what: 'マスタをメンテナンスする画面（検索・一覧・登録・修正・削除）',
    short: 'マスタ保守',
  ),
  PageKinds.search: PageKindWords(
    what: '検索して一覧を見るだけの画面',
    short: '照会（読み取り専用）',
  ),
  PageKinds.detail: PageKindWords(
    what: '1件の内容を読むだけの画面',
    short: '1件の照会',
  ),
  PageKinds.form: PageKindWords(
    what: '1件を入力する画面（新規と編集の両方）',
    short: '1件の入力',
  ),
  PageKinds.wizard: PageKindWords(
    what: '入力をステップに分けた画面',
    short: '段階入力',
  ),
  PageKinds.dashboard: PageKindWords(
    what: '数字とグラフのカードを並べて見せる画面',
    short: '数字とグラフ',
  ),
  PageKinds.report: PageKindWords(
    what: '印刷向けの帳票',
    short: '帳票',
  ),
};

/// Heading word for [kind] (the kind itself when a plugin added it).
String shortWordOf(String kind) => pageKindWords[kind]?.short ?? kind;

/// Full wording for [kind] (empty when a plugin added it — there is no honest
/// sentence to invent for a kind this edition does not know).
String whatWordOf(String kind) => pageKindWords[kind]?.what ?? '';

File: java/src/main/java/io/hatake/core/PageKindWords.java
package io.hatake.core;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * ページ種別の「人に見せる言い方」。
 *
 * <p>1つの種別に2つの言い方を持つ。{@code what} は説明の全文で使う言い方
 * （「検索して一覧に出し、その場で登録・修正・削除までできる画面」）、{@code shortWord} は
 * 1行の要約や画面の索引で使う見出し語（「検索＋一覧＋登録・修正・削除」）。文は表の1行に
 * 入らないし、見出し語は説明にならない。
 *
 * <p><b>正は {@code spec/vocabulary.json}</b>で、ここはその転記（TypeScript 版・Dart 版も
 * 同じ表を転記している）。{@code PageKindWordsTest} が1キーずつ突き合わせるので、手で写して
 * あっても言い方がズレることはない。ズレると CLI と画面で同じ画面の呼び方が変わる。
 *
 * @param what 説明で使う言い方
 * @param shortWord 見出し語（{@code short} は Java の予約語なので名前を変えている）
 */
public record PageKindWords(String what, String shortWord) {

    /** ページ種別 → 言い方（仕様の並び順）。 */
    public static final Map<String, PageKindWords> ALL = all();

    private static Map<String, PageKindWords> all() {
        Map<String, PageKindWords> words = new LinkedHashMap<>();
        words.put("crud", new PageKindWords(
                "検索して一覧に出し、その場で登録・修正・削除までできる画面",
                "検索＋一覧＋登録・修正・削除"));
        words.put("master", new PageKindWords(
                "マスタをメンテナンスする画面（検索・一覧・登録・修正・削除）",
                "マスタ保守"));
        words.put("search", new PageKindWords(
                "検索して一覧を見るだけの画面",
                "照会（読み取り専用）"));
        words.put("detail", new PageKindWords(
                "1件の内容を読むだけの画面",
                "1件の照会"));
        words.put("form", new PageKindWords(
                "1件を入力する画面（新規と編集の両方）",
                "1件の入力"));
        words.put("wizard", new PageKindWords(
                "入力をステップに分けた画面",
                "段階入力"));
        words.put("dashboard", new PageKindWords(
                "数字とグラフのカードを並べて見せる画面",
                "数字とグラフ"));
        words.put("report", new PageKindWords(
                "印刷向けの帳票",
                "帳票"));
        return Map.copyOf(words);
    }

    /** 組み込みのページ種別（{@code page.type} に書ける値）。 */
    public static final List<String> KINDS =
            List.of("crud", "search", "master", "detail", "form", "wizard", "dashboard", "report");

    /** {@code kind} の見出し語。プラグインが足した種別は、その種別名そのもの。 */
    public static String shortOf(String kind) {
        PageKindWords words = ALL.get(kind);
        return words == null ? kind : words.shortWord();
    }

    /**
     * {@code kind} の説明の言い方。知らない種別では空。
     *
     * <p>知らない種別に文を作らないのは、作れば嘘になるため（プラグインの種別が何をする画面
     * かは、この版は知らない）。
     */
    public static String whatOf(String kind) {
        PageKindWords words = ALL.get(kind);
        return words == null ? "" : words.what();
    }
}
